osquery/tables/system/darwin/keychain_utils.cpp

```diff
@@ -72,151 +73,132 @@ std::string getKeychainPath(const SecKeychainItemRef& item) {
+    return "";
   }
 
-  copy = CFNumberCreate(nullptr, kCFNumberIntType, &value);
-  return copy;
+  // BIO_gets() returns amount of data successfully read or written
+  // (if the return value is positive) or that no data was successfully
+  // read or written if the result is 0 or -1.
+  if (BIO_gets(bio_stream, buffer, buffer_size) <= 0) {
+    BIO_free(bio_stream);
+    return "";
+  }
+  BIO_free(bio_stream);
+  return std::string(buffer);
 }
 
-CFDataRef CreatePropertyFromCertificate(const SecCertificateRef& cert,
-                                        const CFTypeRef& oid) {
-  // Set the list of attributes.
-  auto keys = CFArrayCreateMutable(nullptr, 0, &kCFTypeArrayCallBacks);
-  CFArrayAppendValue(keys, oid); // SecCertificateOIDs.h
-
-  // Request dictionary of dictionaries (one for each attribute).
-  auto certificate_values = SecCertificateCopyValues(cert, keys, nullptr);
-  CFRelease(keys);
-  if (certificate_values == nullptr) {
-    return nullptr;
+time_t genEpoch(ASN1_TIME* time) {
+  auto datetime = genHumanReadableDateTime(time);
+  if (datetime.empty()) {
+    return -1;
   }
 
-  if (!CFDictionaryContainsKey(certificate_values, oid)) {
-    // Certificate does not have the requested property.
-    CFRelease(certificate_values);
-    return nullptr;
+  time_t epoch;
+  struct tm tm;
+  // b := abbr month, e := day with leading space instead of leading zero
+  if (strptime(datetime.c_str(), "%b %e %H:%M:%S %Y %Z", &tm) == nullptr) {
+    return -1;
   }
 
-  auto values = (CFDictionaryRef)CFDictionaryGetValue(certificate_values, oid);
-  if (values == nullptr) {
-    CFRelease(certificate_values);
-    return nullptr;
-  }
-  if (!CFDictionaryContainsKey(values, kSecPropertyKeyValue)) {
-    // Odd, there was not value in the property result.
-    CFRelease(certificate_values);
-    return nullptr;
+  // Don't set DST, since strptime() doesn't.
+  // Let mktime() determine whether DST in effect
+  tm.tm_isdst = -1;
+  epoch = mktime(&tm);
+  if (epoch == -1) {
+    return -1;
   }
+  return epoch;
+}
 
-  // Create copy of the property value, which is an index to owned dict.
-  auto property = (CFDataRef)CFDictionaryGetValue(values, kSecPropertyKeyValue);
-  if (CFGetTypeID(property) == CFArrayGetTypeID()) {
-    property = (CFDataRef)CFArrayCreateCopy(nullptr, (CFArrayRef)property);
-  } else if (CFGetTypeID(property) == CFNumberGetTypeID()) {
-    property = (CFDataRef)CFNumberCreateCopy((CFNumberRef)property);
-  } else {
-    property = nullptr;
+// Key Usages (i.e. Digital Signature, CRL Sign etc) in ASN1/OpenSSL
+// are represented as flags. These are then set by doing bitwise OR ops.
+// genKeyUsage() reverses this to figure out which key usages are set.
+std::string genKeyUsage(unsigned long flag) {
+  if (flag == 0) {
+    return "";
   }
-
-  // Release and give the caller control of the property.
-  CFRelease(certificate_values);
-  return property;
+  std::vector<std::string> results;
+  while (flag > 0) {
```

Comment:
I wanted same ordering of key usages as presented in Keychain Access. But now that I think about it, the loss of code readability as a result is not worth it.